uberduck_ml_dev/models/gradtts.py

```diff
@@ -0,0 +1,1152 @@
+        self.dec_dim = hparams.dec_dim
+        self.beta_min = hparams.beta_min
+        self.beta_max = hparams.beta_max
+        self.pe_scale = hparams.pe_scale
+
+        if self.n_spks > 1:
+            self.spk_emb = torch.nn.Embedding(self.n_spks, self.spk_emb_dim)
+        self.encoder = TextEncoder(
+            self.n_vocab,
+            self.n_feats,
+            self.n_enc_channels,
+            self.filter_channels,
+            self.filter_channels_dp,
+            self.n_heads,
+            self.n_enc_layers,
+            self.enc_kernel,
+            self.enc_dropout,
+            self.window_size,
+        )
+        self.decoder = Diffusion(
+            self.n_feats,
+            self.dec_dim,
+            self.n_spks,
+            self.spk_emb_dim,
+            self.beta_min,
+            self.beta_max,
+            self.pe_scale,
+        )
+
+    @property
+    def nparams(self):
+        """
+        Returns number of trainable parameters of the module.
+        """
+        num_params = 0
+        for name, param in self.named_parameters():
+            if param.requires_grad:
+                num_params += np.prod(param.detach().cpu().numpy().shape)
+        return num_params
+
+    def relocate_input(self, x: list):
+        """
+        Relocates provided tensors to the same device set for the module.
+        """
+        device = next(self.parameters()).device
+        for i in range(len(x)):
+            if isinstance(x[i], torch.Tensor) and x[i].device != device:
+                x[i] = x[i].to(device)
+        return x
+
+    @torch.no_grad()
+    def forward(
+        self,
+        x,
+        x_lengths,
+        n_timesteps,
+        temperature=1.0,
+        stoc=False,
+        spk=None,
+        length_scale=1.0,
+    ):
+        """
+        Generates mel-spectrogram from text. Returns:
+            1. encoder outputs
+            2. decoder outputs
+            3. generated alignment
+
+        Args:
+            x (torch.Tensor): batch of texts, converted to a tensor with phoneme embedding ids.
+            x_lengths (torch.Tensor): lengths of texts in batch.
+            n_timesteps (int): number of steps to use for reverse diffusion in decoder.
+            temperature (float, optional): controls variance of terminal distribution.
+            stoc (bool, optional): flag that adds stochastic term to the decoder sampler.
+                Usually, does not provide synthesis improvements.
+            length_scale (float, optional): controls speech pace.
+                Increase value to slow down generated speech and vice versa.
+        """
+        x, x_lengths = self.relocate_input([x, x_lengths])
+
+        if self.n_spks > 1:
+            # Get speaker embedding
+            spk = self.spk_emb(spk)
+
+        # Get encoder_outputs `mu_x` and log-scaled token durations `logw`
+        mu_x, logw, x_mask = self.encoder(x, x_lengths, spk)
+
+        w = torch.exp(logw) * x_mask
```

Comment:
awesome, so looks to me like all we need to do to condition on durations is pass this `w` directly! So excited to try it out!